cpp/ql/lib/semmle/code/cpp/security/InvalidPointerDereference/AllocationToInvalidPointer.qll

```diff
@@ -96,73 +96,111 @@ predicate hasSize(HeuristicAllocationExpr alloc, DataFlow::Node n, int state) {
-      g.comparesLt(left.asOperand(), right.asOperand(), k, true, testIsTrue)
+      // in a guard such as `left <= right + k`.
+      g.comparesLt(left.asOperand(), right.asOperand(), k + 1, true, testIsTrue)
     }
 
     predicate isSink(DataFlow::Node sink) { isSink(_, sink, _, _, _) }
   }
 
   private import DataFlow::Global<SizeBarrierConfig>
 
-  private int getAFlowStateForNode(DataFlow::Node node) {
+  private int getASizeAddend(DataFlow::Node node) {
     exists(DataFlow::Node source |
       flow(source, node) and
       hasSize(_, source, result)
     )
   }
 
+  /**
+   * Holds if `left <= nRight + k` holds if `g` evaluates to `edge`.
+   */
   private predicate operandGuardChecks(
-    IRGuardCondition g, Operand left, Operand right, int state, boolean edge
+    IRGuardCondition g, Operand left, DataFlow::Node right, int k, boolean edge
   ) {
-    exists(DataFlow::Node nLeft, DataFlow::Node nRight, int k |
-      nRight.asOperand() = right and
-      nLeft.asOperand() = left and
-      SizeBarrierConfig::isSink(nLeft, nRight, g, k, edge) and
-      state = getAFlowStateForNode(nRight) and
-      k <= state
+    flowTo(right) and
+    SizeBarrierConfig::isSink(DataFlow::operandNode(left), right, g, k, edge)
+  }
+
+  /**
+   * Gets an instruction `instr` that is guarded by a check such as `instr <= left + delta` where
+   * `left <= _ + k` and `left` is the "small side" of of a relational comparison that checks
+   * whether `left <= size` where `size` is the size of an allocation.
+   */
+  Instruction getABarrierInstruction0(int delta, int k) {
+    exists(IRGuardCondition g, ValueNumber value, Operand left, boolean edge, DataFlow::Node right |
+      // We know:
+      // 1. result <= value + delta (by `bounded`)
+      // 2. value <= right + k (by `operandGuardChecks`).
+      // So:
+      // result <= value + delta (by 1.)
+      //        <= right + k + delta (by 2.)
+      left = value.getAUse() and
+      operandGuardChecks(pragma[only_bind_into](g), pragma[only_bind_into](left), right,
+        pragma[only_bind_into](k), pragma[only_bind_into](edge)) and
+      bounded(result, value.getAnInstruction(), delta) and
+      g.controls(result.getBlock(), edge) and
+      k < getASizeAddend(right)
     )
   }
 
   /**
    * Gets an instruction that is guarded by a guard condition which ensures that
    * the value of the instruction is upper-bounded by size of some allocation.
    */
+  bindingset[state]
+  pragma[inline_late]
   Instruction getABarrierInstruction(int state) {
-    exists(IRGuardCondition g, ValueNumber value, Operand use, boolean edge |
-      use = value.getAUse() and
-      operandGuardChecks(pragma[only_bind_into](g), pragma[only_bind_into](use), _,
-        pragma[only_bind_into](state), pragma[only_bind_into](edge)) and
-      result = value.getAnInstruction() and
-      g.controls(result.getBlock(), edge)
+    exists(int delta, int k |
+      state >= k + delta and
```

Comment:
I guess what I'm asking: what intuition should I have with `state`?

Reply:
`getABarrierInstruction` is called in `pointerAddInstructionHasBounds0` where `state` is given as the delta `d` such that `pai.getRight() <= size + d` and:
- `pai` is the pointer-arithmetic instruction that constructs the out-of-bounds pointer, and
- `size` is the instruction dataflow has determined contains the value of the size of the allocation.

So if we have a program like:
```cpp
int* p = new int[size + 2];
if(n < size + 1) {
  int* end = p + (n + 2);
}
```
`state` would be `2` because the right-hand side of the pointer-arithmetic instruction `p + (size + 2)` is `size + 2`, and `pai.getRight() <= size + 2`. So when we check that `state >= k + delta` we're ensuring that `state` (i.e., the size addend) is greater than the size of the offset we added to the out-of-bounds pointer. In this case `k + delta` is `1` because `k = 1` and `delta = 0`, and since `2 >= 1` the flow is blocked.

Reply:
> Does that make sense?

No. In your example, accessing `end` would be out-of-bounds when `n = size`.

Reply:
Oh no, you're right. Looking at the inline comments that does make sense. `getABarrierInstruction0(delta, k)` gives back an instruction `instr` such that `instr <= "size of allocation" + delta + k`, and when `state >= delta + k` we block flow through instructions `instr` such that `instr <= "size of allocation" + state` (instead of `instr < "size of allocation" + state`). However, when I try to make the inequality `strict` I get a FP from:
```cpp
void test21() {
  int n = 0;
  while (test21_get(n)) n+=2;

  void** xs = new void*[n];

  for (int i = 0; i < n; i += 2) {
    xs[i] = test21_get(i); // GOOD
    xs[i+1] = test21_get(i+1); // GOOD
  }
}
```
I'll investigate this!

Reply:
Aha! Indeed, the `>=` should indeed be a `>`. The FP we got from the above test was because the `boundedImpl` predicate was satisfies for `delta`s that weren't _only_ the most precise ones inferred by the range analysis library. And so we were getting a FP whenever we we managed to pick a not-the-most-precise bound inferred by the library.

- c0a54e90c9453e69cc342755eeac391dfedd2c94 replaces `<=` with `<` and accepts the test regressions, and
- 3d5b1986c9ce0ea50c8461470ea254b429f5f170 fixes those regressions.

Reply:
Oh and for completeness: 2164069eda72f78d0deb9db7bc16d1d3ff6557dd adds the testcase from above.